src/internal/pfsdb/branches.go

```diff
@@ -0,0 +1,73 @@
+package pfsdb
+
+import (
+	"context"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/pfs"
+)
+
+// GetBranch returns a branch by id.
+func GetBranch(ctx context.Context, tx *pachsql.Tx, id BranchID) (*pfs.BranchInfo, error) {
+	branch := new(Branch)
+	if err := tx.GetContext(ctx, branch, `
+		SELECT
+			branch.name,
+			branch.created_at,
+			branch.updated_at,
+			repo.name as "repo.name",
+			repo.type as "repo.type",
+			project.name as "repo.project.name",
+			commit.commit_set_id as "head.commit_set_id",
+			repo.name as "head.repo.name",
+			repo.type as "head.repo.type",
+			project.name as "head.repo.project.name"
+		FROM pfs.branches branch
+			JOIN pfs.repos repo ON branch.repo_id = repo.id
+			JOIN core.projects project ON repo.project_id = project.id
+			JOIN pfs.commits commit ON branch.head = commit.int_id
+		WHERE branch.id = $1
+	`, id); err != nil {
+		return nil, errors.Wrap(err, "could not get branch")
+	}
+	return &pfs.BranchInfo{Branch: branch.Pb(), Head: branch.Head.Pb()}, nil
+}
+
+// UpsertBranch creates a branch if it does not exist, or updates the head commit if it does.
+func UpsertBranch(ctx context.Context, tx *pachsql.Tx, branchInfo *pfs.BranchInfo) (BranchID, error) {
+	if branchInfo.Branch.Repo.Name == "" {
+		return 0, errors.Errorf("repo name required")
+	}
+	if branchInfo.Branch.Repo.Type == "" {
+		return 0, errors.Errorf("repo type required")
+	}
+	if branchInfo.Branch.Repo.Project.Name == "" {
+		return 0, errors.Errorf("project name required")
+	}
+	if branchInfo.Head.Id == "" {
+		return 0, errors.Errorf("head commit required")
+	}
+
```

Comment:
[nit] we tend not to like inserting new lines to separate code blocks throughout pfs code.